python/feldera/sql_context.py

```diff
@@ -0,0 +1,278 @@
+
+        self.build_mode = BuildMode.CREATE
+        return self
+
+    def get(self) -> Self:
+        """
+        Sets the build mode to GET, meaning that an existing pipeline will be used.
+        """
+
+        self.build_mode = BuildMode.GET
+        return self
+
+    def get_or_create(self) -> Self:
+        """
+        Sets the build mode to GET_OR_CREATE, meaning that an existing pipeline will be used if it exists,
+        else a new one will be created.
+        """
+
+        self.build_mode = BuildMode.GET_OR_CREATE
+        return self
+
+    def register_table(self, table_name: str, schema: Optional[SQLSchema] = None, ddl: str = None):
+        """
+        Registers a table with the SQLContext. The table can be registered with a schema or with the SQL DDL.
+        One of the two must be provided, but not both.
+        Auto inserts the trailing semicolon if not present.
+        In the future, schema will be inferred from the data provided from applicable sources.
+
+        :param table_name: The name of the table.
+        :param schema: The schema of the table.
+        :param ddl: The SQL DDL of the table.
+        """
+
+        if not schema and not ddl:
+            raise ValueError("Schema inference isn't supported yet, either provide a schema or the SQL DDL")
+
+        if schema and ddl:
+            raise ValueError("Provide either a schema or the SQL DDL, not both")
+
+        if ddl:
+            self.register_table_from_sql(ddl)
+            return
+
+        if schema:
+            self.tables[table_name] = SQLTable(table_name, schema=schema)
+        else:
+            self.todo_tables[table_name] = None
+
+    def register_table_from_sql(self, ddl: str):
+        """
+        Registers a table with the provided SQL DDL.
+        Auto inserts the trailing semicolon if not present.
+
+        :param ddl: The SQL DDL of the table.
+        """
+
+        if ddl[-1] != ';':
+            ddl += ';'
+
+        name = _table_name_from_sql(ddl)
+
+        self.tables[name] = SQLTable(name, ddl)
+
+    def input_from_pandas(self, table_name: str, df: pandas.DataFrame):
+        """
+        Adds a pandas DataFrame to the input buffer of the SQLContext, to be pushed to the pipeline.
+
+        :param table_name: The name of the table.
+        :param df: The pandas DataFrame to be pushed to the pipeline.
+        """
+
+        tbl = self.tables.get(table_name)
+
+        if tbl:
+            # tbl.validate_schema(df)   TODO: something like this would be nice
+            self.http_input_buffer.append({tbl.name: df.to_dict('records')})
+            return
+
+        tbl = self.todo_tables.get(table_name)
+
+        if not tbl:
+            raise ValueError(f"Table {table_name} not registered")
+
+        # tbl.infer_schema(df)  TODO: support schema inference
+
+        self.tables[table_name] = tbl
+        self.todo_tables.pop(table_name)
+
+        self.http_input_buffer.append({tbl.name: df.to_dict('records')})
```

Comment:
I wonder if a better API would store a reference to the dataframe and would only read it after starting the pipeline. This would be more inline with other connectors that only start running with the pipeline, plus we wouldn't need to buffer potentially large data chunks in memory.

Reply:
So something like the following? 

```py
input_buffer.append({tbl.name: df}) # in input_from_pandas

# in run_to_completion
for tbl_name, df in input_buffer.items():
   sql.client.push_to_pipeline(..., table_name, data.to_dict('records'))
```

Reply:
yes, but I think we decided not to do this for now?

Reply:
Yes.